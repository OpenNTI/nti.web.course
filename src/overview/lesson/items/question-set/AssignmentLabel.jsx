import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import {List, DateTime} from 'nti-web-commons';
import {scoped} from 'nti-lib-locale';

const DEFAULT_TEXT = {
	draft: 'Draft',
	maxTime: '%(maxTime)s time limit',
	completed: {
		label: 'completed ',
		graded: 'graded ',
		submittedAt: 'Submitted At %(date)s',
		overTime: 'overtime',
		overDue: 'overdue',
		overTimeTip: '%(time)s overtime',
		overDueTip: '%(time)s overdue',
		modOpen: '(',
		modClose: ') ',
		modSeparator: ', '
	},
	due: {
		today: 'Due Today at %(time)s',
		due: 'Due %(date)s',
		availableNow: 'Available Now',
		available: 'Available %(date)s'
	},
	excused: {
		label: 'Excused Grade',
		tip: 'This assignment will NOT count towards your grade'
	},
	required: 'Required'
};
const t = scoped('course.overview.lesson.overview.question-set.AssignmentLabel', DEFAULT_TEXT);

const getNaturalDuration = (...args) => DateTime.getNaturalDuration(...args);
const formatDate = (...args) => DateTime.format(...args);
const isToday = (...args) => DateTime.isToday(...args);

export default class LessonOverviewAssignmentLabel extends React.Component {
	static propTypes = {
		assignment: PropTypes.object,
		assignmentHistory: PropTypes.object,
		required: PropTypes.bool
	}

	state = {}

	componentWillReceiveProps (nextProps) {
		const {assignment:newAssignment, assignmentHistory:newHistory} = nextProps;
		const {assignment:oldAssignment, assignmentHistory:oldHistory} = this.props;

		if (newAssignment !== oldAssignment || newHistory !== oldHistory) {
			this.setupFor(nextProps);
		}
	}


	componentDidMount () {
		this.setupFor(this.props);
	}


	setupFor (props = this.props) {
		const {assignment: a, assignmentHistory: h} = props;
		const canEdit = a.hasLink('date-edit') || a.hasLink('edit');

		const state = {
			now: new Date(),
			availableDate: a && a.getAssignedDate(),
			dueDate: a && a.getDueDate(),
			isNoSubmit: (a && a.isNonSubmit()) || (h && h.isSyntheticSubmission()),
			isSynthetic: h && h.isSyntheticSubmission(),
			isDraft: a && !a.isPublished(),
			isTimed: a && a.isTimed,
			maxTime: a && a.isTimed && a.getMaximumTimeAllowed && a.getMaximumTimeAllowed(),
			duration: a && a.isTimed && a.getDuration && a.getDuration(),

			isSubmitted: h && h.isSubmitted(),
			completedDate: !canEdit && h && h.Submission && h.Submission.getCreatedTime(),
			isExcused: h && h.grade && h.grade.isExcused()

		};

		this.setState(state);
	}



	render () {
		return (
			<List.SeparatedInline className="lesson-overview-assignment-status">
				{this.renderDraft()}
				{this.renderTimed()}
				{this.renderCompletion()}
				{this.renderRequired()}
				{this.renderDue()}
				{this.renderExcused()}
			</List.SeparatedInline>
		);
	}


	renderDraft () {
		const {isDraft} = this.state;

		return isDraft ?
			(<span className="draft">{t('draft')}</span>) :
			null;
	}


	renderTimed () {
		const {duration, isSynthetic} = this.state;

		if (isSynthetic) { return null; }

		return duration ?
			this.renderDuration() :
			this.renderMaxTime();
	}

	renderDuration () {
		const {maxTime, duration} = this.state;
		const ontime = duration <= maxTime;
		const cls = cx('duration', {ontime, overtime: !ontime});

		return (
			<span className={cls}>{getNaturalDuration(duration, 2)}</span>
		);
	}

	renderMaxTime () {
		const {maxTime} = this.state;

		return maxTime ?
			(<span className="max-time">{t('maxTime', {maxTime: getNaturalDuration(maxTime, 2, true)})}</span>) :
			null;
	}


	renderCompletion () {
		const {completedDate, isNoSubmit, dueDate, maxTime, duration} = this.state;

		if (!completedDate) { return null; }

		const late = dueDate && completedDate >= dueDate;
		const qtip = t('completed.submittedAt', {date: formatDate(completedDate, 'h:mm A M/D/YYYY')});

		const overtime = maxTime && duration && duration > maxTime ? t('completed.overTimeTip', {time: getNaturalDuration(duration - maxTime, 1)}) : null;
		const overdue = late ? t('completed.overDueTip', {time: getNaturalDuration(completedDate.getTime() - dueDate.getTime())}) : null;

		return (
			<span className={cx('completed', {late, ontime: !late})}>
				<span className="label" data-qtip={qtip}>{isNoSubmit ? t('completed.graded') : t('completed.label')}</span>
				{(overtime != null || overdue != null) && (
					<React.Fragment>
						<span className="mod-open errors">{t('completed.modOpen')}</span>
						{overtime != null && (<span className="errors" data-qtip={overtime}>{t('completed.overTime')}</span>)}
						{overtime != null && overdue != null && (<span className="mod-separator errors">{t('completed.modSeparator')}</span>)}
						{overdue != null && (<span className="errors" data-qtip={overdue}>{t('completed.overDue')}</span>)}
						<span className="mod-close erros">{t('completed.modClose')}</span>
					</React.Fragment>
				)}
				<DateTime className="date" date={completedDate} format="dddd, MMMM D" />
			</span>
		);
	}


	renderDue () {
		const {completedDate, dueDate, availableDate, now, isDraft, isNoSubmit} = this.state;

		if (completedDate) { return null; }

		const format = date => formatDate(date, 'dddd, MMMM D h:mm A z');

		const dueToday = dueDate > now && isToday(dueDate);
		const late = !isNoSubmit && dueDate && dueDate <= now;

		let text = '';

		if (dueToday) {
			text = t('due.today', {time: formatDate(dueDate, 'h:m a z')});
		} else if (availableDate > now && dueDate > now) {
			text = t('due.available', {date: format(availableDate)});
		} else if (dueDate) {
			text = t('due.due', {date: format(dueDate)});
		} else if (availableDate < now) {
			text = t('due.availableNow');
		} else if (!isDraft && availableDate) {
			text = t('due.available', {date: format(availableDate)});
		}

		return !text ? null : (
			<span className={cx('due', {today: dueToday, late})}>
				{text}
			</span>
		);
	}


	renderRequired () {
		const {required} = this.props;

		return required && (<div className="required">{t('required')}</div>);
	}


	renderExcused () {
		const {isExcused} = this.state;

		return !isExcused ?
			null :
			(<span className="excused" data-qtip={t('excused.tip')}>{t('excused.label')}</span>);
	}
}
