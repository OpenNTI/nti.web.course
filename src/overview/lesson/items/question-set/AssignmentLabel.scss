@import '~@nti/style-common';

.lesson-overview-assignment-status.nti-separated-inline-list {
	font: normal 600 0.625rem/1.3 var(--body-font-family);
	text-transform: uppercase;
	color: var(--tertiary-grey);

	li {
		display: inline-block;
	}

	.completed {
		&.ontime:not(.submission-failed) .status-label {
			color: var(--secondary-green);
			background: none;
			padding-left: 0;
		}

		.errors {
			color: var(--primary-red);
		}
	}

	.draft {
		margin-right: 3px;
	}

	.due {
		&.today {
			color: var(--primary-orange);
		}

		&.late {
			color: var(--primary-red);
		}

		i {
			margin-left: 3px;
		}

		&.editable .label:hover {
			text-decoration: underline;
			border-radius: 10px;
			cursor: pointer;
		}
	}

	.excused {
		color: var(--primary-red);
	}

	.failed {
		color: var(--primary-red);
	}
}
