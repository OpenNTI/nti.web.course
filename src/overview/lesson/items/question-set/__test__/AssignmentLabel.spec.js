import React from 'react';
import renderer from 'react-test-renderer';

import AssignmentLabel from '../AssignmentLabel';

function buildAssignment(
	assignedDate,
	dueDate,
	lastModified,
	maxTimeAllowed,
	duration,
	isPublished,
	isSubmitted,
	isTimed,
	noSubmit,
	canEdit
) {
	return {
		getAssignedDate: () =>
			assignedDate && typeof assignedDate === 'string'
				? new Date(assignedDate)
				: assignedDate,
		getDueDate: () =>
			dueDate && typeof dueDate === 'string'
				? new Date(dueDate)
				: dueDate,
		getLastModified: () =>
			lastModified && typeof lastModified === 'string'
				? new Date(lastModified)
				: lastModified,
		isPublished: () => isPublished,
		isNonSubmit: () => noSubmit,
		hasLink: () => canEdit,
		getMaximumTimeAllowed: () => maxTimeAllowed,
		getDuration: () => duration,
		isSubmitted: () => isSubmitted,
		isTimed,
	};
}

function assignmentBuilder() {
	return {
		assignedDateVal: null,
		dueDateVal: null,
		lastModifiedVal: new Date(),
		maxTimeAllowedVal: null,
		durationVal: null,
		isPublishedVal: false,
		isTimedVal: false,
		isSubmittedVal: false,
		noSubmitVal: false,
		canEditVal: false,

		assignedDate: function (val) {
			this.assignedDateVal = val;
			return this;
		},
		dueDate: function (val) {
			this.dueDateVal = val;
			return this;
		},
		lastModified: function (val) {
			this.lastModifiedVal = val;
			return this;
		},
		maxTimeAllowed: function (val) {
			this.maxTimeAllowedVal = val;
			return this;
		},
		duration: function (val) {
			this.durationVal = val;
			return this;
		},
		isPublished: function () {
			this.isPublishedVal = true;
			return this;
		},
		isSubmitted: function () {
			this.isSubmittedVal = true;
			return this;
		},
		isTimed: function () {
			this.isTimedVal = true;
			return this;
		},
		noSubmit: function () {
			this.noSubmitVal = true;
			return this;
		},
		canEdit: function () {
			this.canEditVal = true;
			return this;
		},
		build: function () {
			return buildAssignment(
				this.assignedDateVal,
				this.dueDateVal,
				this.lastModifiedVal,
				this.maxTimeAllowedVal,
				this.durationVal,
				this.isPublishedVal,
				this.isSubmittedVal,
				this.isTimedVal,
				this.noSubmitVal,
				this.canEditVal
			);
		},
	};
}

function buildAssignmentHistory(
	completedDate,
	isSynthetic,
	isSubmitted,
	isExcused
) {
	return {
		isSyntheticSubmission: () => isSynthetic,
		isSubmitted: () => isSubmitted,
		Submission: {
			getCreatedTime: () =>
				completedDate && typeof completedDate === 'string'
					? new Date(completedDate)
					: null,
		},
		grade: {
			isExcused: () => isExcused,
		},
	};
}

/* eslint-env jest */
describe('Course overview assignment label test', () => {
	test('Completed but late', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.dueDate('10/22/2017')
			.isPublished()
			.isSubmitted()
			.noSubmit()
			.build();

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					'10/25/2017', // completedDate
					false, // isSynthetic
					true, // isSubmitted
					false // isExcused
				)}
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Draft', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.build();

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					null, // completedDate
					false, // isSynthetic
					true, // isSubmitted
					false // isExcused
				)}
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('With duration', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.isPublished()
			.isTimed()
			.duration(1000000)
			.build();

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					null, // completedDate
					false, // isSynthetic
					true, // isSubmitted
					false // isExcused
				)}
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Completed on time', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.dueDate('10/30/2017')
			.isPublished()
			.isSubmitted()
			.noSubmit()
			.build();

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					'10/25/2017', // completedDate
					false, // isSynthetic
					true, // isSubmitted
					false // isExcused
				)}
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Not completed, overdue', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.dueDate('10/30/2017')
			.isPublished()
			.build();

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					null, // completedDate
					false, // isSynthetic
					false, // isSubmitted
					false // isExcused
				)}
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Not completed, overdue, but is no submit assignment', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.dueDate('10/30/2017')
			.isPublished()
			.noSubmit()
			.build();

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					null, // completedDate
					false, // isSynthetic
					false, // isSubmitted
					false // isExcused
				)}
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Not completed, but excused', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.dueDate('10/30/2017')
			.isPublished()
			.build();

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					null, // completedDate
					false, // isSynthetic
					false, // isSubmitted
					true // isExcused
				)}
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Completable, optional, editable', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.dueDate('10/30/2017')
			.isPublished()
			.canEdit()
			.build();

		const overviewItemRef = {
			isCompletable: () => true,
		};

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					null, // completedDate
					false, // isSynthetic
					false, // isSubmitted
					false // isExcused
				)}
				overviewItemRef={overviewItemRef}
				onRequirementChange={() => {}}
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Completable, required, not editable', async () => {
		const assignment = assignmentBuilder()
			.assignedDate('10/22/2017')
			.dueDate('10/30/2017')
			.isPublished()
			.build();

		const cmp = renderer.create(
			<AssignmentLabel
				assignment={assignment}
				assignmentHistory={buildAssignmentHistory(
					null, // completedDate
					false, // isSynthetic
					false, // isSubmitted
					false // isExcused
				)}
				required
			/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
