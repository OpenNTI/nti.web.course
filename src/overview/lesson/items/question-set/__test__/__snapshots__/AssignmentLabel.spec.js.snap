// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Course overview assignment label test Completable, optional, editable 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <div
      aria-haspopup="dialog"
      className="flyout-trigger require-control-value require-control flyout-closed"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      tabIndex={0}
    >
      <span>
        Optional
      </span>
      <i
        className="icon-chevron-down"
      />
    </div>
  </li>
  <li>
    <span
      className="attempts"
    >
      1 Attempt
    </span>
  </li>
  <li>
    <span
      className="due late"
    >
      Due Monday, October 30, 12:00 AM UTC
    </span>
  </li>
</ul>
`;

exports[`Course overview assignment label test Completable, required, not editable 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <div
      className="lesson-overview-item-required-label"
    >
      Required
    </div>
  </li>
  <li>
    <span
      className="attempts"
    >
      1 Attempt
    </span>
  </li>
  <li>
    <span
      className="due late"
    >
      Due Monday, October 30, 12:00 AM UTC
    </span>
  </li>
</ul>
`;

exports[`Course overview assignment label test Completed but late 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <span
      className="completed late"
    >
      <span
        className="status-label"
        data-qtip="Submitted At 12:00 AM 10/25/2017"
      >
        graded 
      </span>
      <time
        className="date ntiText"
        dateTime="2017-10-25T00:00:00Z"
      >
        Wednesday, October 25
      </time>
    </span>
  </li>
</ul>
`;

exports[`Course overview assignment label test Completed on time 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <span
      className="completed ontime"
    >
      <span
        className="status-label"
        data-qtip="Submitted At 12:00 AM 10/25/2017"
      >
        graded 
      </span>
      <time
        className="date ntiText"
        dateTime="2017-10-25T00:00:00Z"
      >
        Wednesday, October 25
      </time>
    </span>
  </li>
</ul>
`;

exports[`Course overview assignment label test Draft 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <span
      className="draft"
    >
      Draft
    </span>
  </li>
</ul>
`;

exports[`Course overview assignment label test Not completed, but excused 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <span
      className="attempts"
    >
      1 Attempt
    </span>
  </li>
  <li>
    <span
      className="due late"
    >
      Due Monday, October 30, 12:00 AM UTC
    </span>
  </li>
  <li>
    <span
      className="excused"
      data-qtip="This assignment will NOT count towards your grade"
    >
      Excused Grade
    </span>
  </li>
</ul>
`;

exports[`Course overview assignment label test Not completed, overdue 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <span
      className="attempts"
    >
      1 Attempt
    </span>
  </li>
  <li>
    <span
      className="due late"
    >
      Due Monday, October 30, 12:00 AM UTC
    </span>
  </li>
</ul>
`;

exports[`Course overview assignment label test Not completed, overdue, but is no submit assignment 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <span
      className="attempts"
    >
      1 Attempt
    </span>
  </li>
  <li>
    <span
      className="due"
    >
      Due Monday, October 30, 12:00 AM UTC
    </span>
  </li>
</ul>
`;

exports[`Course overview assignment label test With duration 1`] = `
<ul
  className="nti-separated-inline-list lesson-overview-assignment-status"
>
  <li>
    <span
      className="duration overtime"
    >
      16 Minutes, 40 Seconds
    </span>
  </li>
  <li>
    <span
      className="due"
    >
      Available Now
    </span>
  </li>
</ul>
`;
